[Concurrency] Introduce a small executable test using asynchronous futures

DougGregor opened the pull request:
Rather than immediately running the task synchronously within
runDetached, return the handle to the newly-created task. Add a method
task.Handle.run() to execute the task. This is just a temporary hack
that should not persist in the API, but it lets us launch tasks on a
particular Dispatch queue:

```swift
extension DispatchQueue {
  func async<R>(execute: @escaping () async -> R) -> Task.Handle<R> {
    let handle = Task.runDetached(operation: execute)

    // Run the task
    _ = { self.async { handle.run() } }()

    return handle
  }
}
```

One can pass asynchronous work to DispatchQueue.async, which will
schedule that work on the dispatch queue and return a handle. Another
asynchronous task can then read the result.

Yay for rdar://71125519.

DougGregor committed:
[Concurrency] Stop runDetached from actually printing/running the task.

Rather than immediately running the task synchronously within
runDetached, return the handle to the newly-created task. Add a method
task.Handle.run() to execute the task. This is just a temporary hack
that should not persist in the API, but it lets us launch tasks on a
particular Dispatch queue:

```swift
extension DispatchQueue {
  func async<R>(execute: @escaping () async -> R) -> Task.Handle<R> {
    let handle = Task.runDetached(operation: execute)

    // Run the task
    _ = { self.async { handle.run() } }()

    return handle
  }
}
```

One can pass asynchronous work to DispatchQueue.async, which will
schedule that work on the dispatch queue and return a handle. Another
asynchronous task can then read the result.

Yay for rdar://71125519.

DougGregor commented:
@swift-ci please test

ktoso reviewed (approved):
Thanks, helped me kickstart work on the groups 👍 

Your call if we want to merge or not, might as well I suppose.

// failure seems unrelated https://ci.swift.org/job/swift-PR-osx/24459/testReport/Swift(iphonesimulator-i386)/IRGen/objc_generic_class_stub_swift/

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test

DougGregor committed:
[Concurrency] Add an async call into our async task for basic futures test.

Extend the basic futures test to include a call to an async function so
that we exercise coroutine splitting.

DougGregor commented:
@swift-ci please smoke test
